# Issue 28842: Uncaught AttributeError in integrate()

archive/issues_028605.json:
```json
{
    "body": "As reported in this [sage-devel post](https://groups.google.com/d/msg/sage-devel/_wwyqHmE5Lg/M24X8Mv_AAAJ), some integral computations may go through SymPy and return a `RootSum` object, which has currently no translation to Sage (cf. #16816)\n\n```\nsage: integrate(1/(x^4 + x^3 + 1), x)\n...\nAttributeError: 'RootSum' object has no attribute '_sage_'\n```\nUntil #16816 is solved, the current ticket proposes a quick fix by catching the `AttributeError` as if no valid result were found, so that we get \n\n```\nsage: integrate(1/(x^4 + x^3 + 1), x)\nintegrate(1/(x^4 + x^3 + 1), x)\n```\nwhich was the behavior prior to Sage 8.9, i.e. prior to the automatic use of SymPy if Maxima fails.\n\n**CC:**  @dimpase @EmmanuelCharpentier\n\n**Keywords:** integrate, integral, sympy, RootSum\n\n**Branch/Commit:** [e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)\n\n**Reviewer:** Fr\u00e9d\u00e9ric Chapoton\n\n**Author:** Eric Gourgoulhon\n\nIssue created by migration from https://trac.sagemath.org/ticket/28842\n\n",
    "closed_at": "2019-12-09T22:54:28Z",
    "created_at": "2019-12-04T21:57:39Z",
    "labels": [
        "component: symbolics",
        "bug"
    ],
    "milestone": "https://github.com/sagemath/sage/milestones/sage-9.0",
    "title": "Uncaught AttributeError in integrate()",
    "type": "issue",
    "url": "https://github.com/sagemath/sage/issues/28842",
    "user": "https://github.com/egourgoulhon"
}
```
As reported in this [sage-devel post](https://groups.google.com/d/msg/sage-devel/_wwyqHmE5Lg/M24X8Mv_AAAJ), some integral computations may go through SymPy and return a `RootSum` object, which has currently no translation to Sage (cf. #16816)

```
sage: integrate(1/(x^4 + x^3 + 1), x)
...
AttributeError: 'RootSum' object has no attribute '_sage_'
```
Until #16816 is solved, the current ticket proposes a quick fix by catching the `AttributeError` as if no valid result were found, so that we get 

```
sage: integrate(1/(x^4 + x^3 + 1), x)
integrate(1/(x^4 + x^3 + 1), x)
```
which was the behavior prior to Sage 8.9, i.e. prior to the automatic use of SymPy if Maxima fails.

**CC:**  @dimpase @EmmanuelCharpentier

**Keywords:** integrate, integral, sympy, RootSum

**Branch/Commit:** [e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)

**Reviewer:** Frédéric Chapoton

**Author:** Eric Gourgoulhon

Issue created by migration from https://trac.sagemath.org/ticket/28842





---

archive/issue_comments_452699.json:
```json
{
    "body": "**Commit:** [14c75619f8ff530c2887bc3be109aa670ad0b6f1](https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1)",
    "created_at": "2019-12-04T22:08:38Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452699",
    "user": "https://github.com/egourgoulhon"
}
```

**Commit:** [14c75619f8ff530c2887bc3be109aa670ad0b6f1](https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1)



---

archive/issue_comments_452700.json:
```json
{
    "body": "<a id='comment:1'></a>\n**New commits:**\n<table><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1\">14c7561</a></td><td><code>#28842: catch AttributeError from SymPy's RootSum</code></td></tr></table>\n",
    "created_at": "2019-12-04T22:08:38Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452700",
    "user": "https://github.com/egourgoulhon"
}
```

<a id='comment:1'></a>
**New commits:**
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1">14c7561</a></td><td><code>#28842: catch AttributeError from SymPy's RootSum</code></td></tr></table>




---

archive/issue_comments_452701.json:
```json
{
    "body": "**Branch:** [public/symbolics/integrate_RootSum_error-28842](https://github.com/sagemath/sagetrac-mirror/tree/public/symbolics/integrate_RootSum_error-28842)",
    "created_at": "2019-12-04T22:08:38Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452701",
    "user": "https://github.com/egourgoulhon"
}
```

**Branch:** [public/symbolics/integrate_RootSum_error-28842](https://github.com/sagemath/sagetrac-mirror/tree/public/symbolics/integrate_RootSum_error-28842)



---

archive/issue_events_258261.json:
```json
{
    "actor": "https://github.com/egourgoulhon",
    "created_at": "2019-12-04T22:08:38Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258261"
}
```



---

archive/issue_comments_452702.json:
```json
{
    "body": "**Author:** Eric Gourgoulhon",
    "created_at": "2019-12-04T22:08:38Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452702",
    "user": "https://github.com/egourgoulhon"
}
```

**Author:** Eric Gourgoulhon



---

archive/issue_comments_452703.json:
```json
{
    "body": "<a id='comment:2'></a>\nIf/when patchbot says this is good, I think this is a good interim solution.  Good catch.",
    "created_at": "2019-12-05T04:38:34Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452703",
    "user": "https://github.com/kcrisman"
}
```

<a id='comment:2'></a>
If/when patchbot says this is good, I think this is a good interim solution.  Good catch.



---

archive/issue_comments_452704.json:
```json
{
    "body": "<a id='comment:3'></a>\npatchbot says\n\n```\n++        Check that trac:`28842` is fixed::\n+Bad trac link inserted on 1 non-empty lines\n```",
    "created_at": "2019-12-05T07:23:48Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452704",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:3'></a>
patchbot says

```
++        Check that trac:`28842` is fixed::
+Bad trac link inserted on 1 non-empty lines
```



---

archive/issue_comments_452705.json:
```json
{
    "body": "<a id='comment:4'></a>\n**Branch pushed to git repo; I updated commit sha1.** **New commits:**\n<table><tr><td><a href=\"https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98\">e5a3f32</a></td><td><code>#28842: corrected trac link</code></td></tr></table>\n",
    "created_at": "2019-12-05T08:27:10Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452705",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

<a id='comment:4'></a>
**Branch pushed to git repo; I updated commit sha1.** **New commits:**
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98">e5a3f32</a></td><td><code>#28842: corrected trac link</code></td></tr></table>




---

archive/issue_comments_452706.json:
```json
{
    "body": "**Changing commit** from \"[14c75619f8ff530c2887bc3be109aa670ad0b6f1](https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1)\" to \"[e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)\".",
    "created_at": "2019-12-05T08:27:10Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452706",
    "user": "https://trac.sagemath.org/admin/accounts/users/git"
}
```

**Changing commit** from "[14c75619f8ff530c2887bc3be109aa670ad0b6f1](https://github.com/sagemath/sagetrac-mirror/commit/14c75619f8ff530c2887bc3be109aa670ad0b6f1)" to "[e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)".



---

archive/issue_comments_452707.json:
```json
{
    "body": "<a id='comment:5'></a>\nReplying to [@fchapoton](#comment%3A3):\n> patchbot says\n> \n> ```\n> ++        Check that trac:`28842` is fixed::\n> +Bad trac link inserted on 1 non-empty lines\n> ```\n\nOups! This is corrected in the above commit.",
    "created_at": "2019-12-05T08:27:56Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452707",
    "user": "https://github.com/egourgoulhon"
}
```

<a id='comment:5'></a>
Replying to [@fchapoton](#comment%3A3):
> patchbot says
> 
> ```
> ++        Check that trac:`28842` is fixed::
> +Bad trac link inserted on 1 non-empty lines
> ```

Oups! This is corrected in the above commit.



---

archive/issue_events_258262.json:
```json
{
    "actor": "https://github.com/embray",
    "created_at": "2019-12-05T12:06:05Z",
    "event": "renamed",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "rename": {
        "from": "Uncatched AttributeError in integrate()",
        "to": "Uncaught AttributeError in integrate()"
    },
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258262"
}
```



---

archive/issue_comments_452708.json:
```json
{
    "body": "<a id='comment:6'></a>\nMinor English nitpick :)",
    "created_at": "2019-12-05T12:06:05Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452708",
    "user": "https://github.com/embray"
}
```

<a id='comment:6'></a>
Minor English nitpick :)



---

archive/issue_events_258263.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2019-12-05T12:11:45Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "label": "needs review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258263"
}
```



---

archive/issue_events_258264.json:
```json
{
    "actor": "https://github.com/fchapoton",
    "created_at": "2019-12-05T12:11:45Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258264"
}
```



---

archive/issue_comments_452709.json:
```json
{
    "body": "<a id='comment:7'></a>\nok, let it be",
    "created_at": "2019-12-05T12:11:45Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452709",
    "user": "https://github.com/fchapoton"
}
```

<a id='comment:7'></a>
ok, let it be



---

archive/issue_comments_452710.json:
```json
{
    "body": "**Reviewer:** Fr\u00e9d\u00e9ric Chapoton",
    "created_at": "2019-12-05T12:11:45Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452710",
    "user": "https://github.com/fchapoton"
}
```

**Reviewer:** Frédéric Chapoton



---

archive/issue_comments_452711.json:
```json
{
    "body": "<a id='comment:8'></a>\nReplying to [@embray](#comment%3A6):\n> Minor English nitpick :)\n\nShame on me! Thanks for correcting it.",
    "created_at": "2019-12-05T12:21:31Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452711",
    "user": "https://github.com/egourgoulhon"
}
```

<a id='comment:8'></a>
Replying to [@embray](#comment%3A6):
> Minor English nitpick :)

Shame on me! Thanks for correcting it.



---

archive/issue_comments_452712.json:
```json
{
    "body": "<a id='comment:9'></a>\nReplying to [@fchapoton](#comment%3A7):\n> ok, let it be\n\nThanks for the review!",
    "created_at": "2019-12-05T12:22:37Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452712",
    "user": "https://github.com/egourgoulhon"
}
```

<a id='comment:9'></a>
Replying to [@fchapoton](#comment%3A7):
> ok, let it be

Thanks for the review!



---

archive/issue_comments_452713.json:
```json
{
    "body": "<a id='comment:10'></a>\n> > Minor English nitpick :)\n\n> Shame on me! Thanks for correcting it.\n\nHaha, I thought about doing this but since it was just in the Trac title I didn't bother.\n\n> patchbot says\n\nI was wondering if that was the correct syntax.  Glad patchbot saw it.",
    "created_at": "2019-12-05T14:38:00Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452713",
    "user": "https://github.com/kcrisman"
}
```

<a id='comment:10'></a>
> > Minor English nitpick :)

> Shame on me! Thanks for correcting it.

Haha, I thought about doing this but since it was just in the Trac title I didn't bother.

> patchbot says

I was wondering if that was the correct syntax.  Glad patchbot saw it.



---

archive/issue_events_258265.json:
```json
{
    "actor": "https://github.com/vbraun",
    "created_at": "2019-12-09T22:54:28Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "label": "positive review",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258265"
}
```



---

archive/issue_events_258266.json:
```json
{
    "actor": "https://github.com/vbraun",
    "commit_id": "4fb8ce65786764fb38331a588f7d8c5fe8e1af4b",
    "created_at": "2019-12-09T22:54:28Z",
    "event": "closed",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/28842#event-258266"
}
```



---

archive/issue_comments_452714.json:
```json
{
    "body": "**Changing branch** from \"[public/symbolics/integrate_RootSum_error-28842](https://github.com/sagemath/sagetrac-mirror/tree/public/symbolics/integrate_RootSum_error-28842)\" to \"[e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)\".",
    "created_at": "2019-12-09T22:54:28Z",
    "issue": "https://github.com/sagemath/sage/issues/28842",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/28842#issuecomment-452714",
    "user": "https://github.com/vbraun"
}
```

**Changing branch** from "[public/symbolics/integrate_RootSum_error-28842](https://github.com/sagemath/sagetrac-mirror/tree/public/symbolics/integrate_RootSum_error-28842)" to "[e5a3f32c6cb2d62384eadf132cb182e25ef01b98](https://github.com/sagemath/sagetrac-mirror/commit/e5a3f32c6cb2d62384eadf132cb182e25ef01b98)".
